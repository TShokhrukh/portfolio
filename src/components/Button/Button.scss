
.Button {
  cursor: pointer;
  border: 1px solid $purple;
  padding: $spacer * 1.25 $spacer * 2;
  border-radius: $border-radius;
  background: none;
  color: $purple;
  display: flex;
  align-items: center;
}
