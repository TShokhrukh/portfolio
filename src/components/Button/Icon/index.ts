export { ButtonIcon as default } from './Button-Icon'
