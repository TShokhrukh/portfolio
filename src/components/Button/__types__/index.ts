export * from './Button.types'
export * from './Button_variant_link.types'
export * from './Button_disabled.types'
export * from './Button_rounded.types'
export * from './Button-Icon.types'
