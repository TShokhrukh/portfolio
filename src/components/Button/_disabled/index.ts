export * from './Button_disabled'
