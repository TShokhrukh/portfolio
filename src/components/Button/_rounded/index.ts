export * from './Button_rounded'
