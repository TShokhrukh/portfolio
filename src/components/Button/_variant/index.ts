export * from './Button_variant_link'
