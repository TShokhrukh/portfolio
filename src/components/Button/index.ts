export * from './build'
export * from './__types__'
export * from './_disabled'
export * from './_rounded'
export * from './_variant'
export { Button as ButtonBase } from './Button'
