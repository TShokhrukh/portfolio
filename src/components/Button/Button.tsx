import React, { FC } from 'react'

import { ButtonIcon } from './Icon/Button-Icon'
import { cnButton } from './Button.const'
import type { IButtonProps, TRef } from './__types__'

import './Button.scss'

export const Button: FC<IButtonProps> = ({
  component: Component = 'button',
  type = 'button',
  children,
  className,
  startIcon,
  endIcon,
  innerRef,
  ...props
}) => {
  let ref: TRef = { ref: innerRef }

  if (typeof Component === 'function') {
    ref = { innerRef }
  }

  return (
    <Component {...props} {...(ref)}
      className={cnButton({}, [className])}
      type={type}>
      {startIcon && <ButtonIcon icon={startIcon} direction='start' />}
      {children}
      {endIcon && <ButtonIcon icon={endIcon} direction='end' />}
    </Component>
  )
}
