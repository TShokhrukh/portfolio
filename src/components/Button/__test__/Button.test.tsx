/* global test, jest, expect, describe */
import React from 'react'
import { fireEvent, render, screen } from '@testing-library/react'
import { Button } from '../Button'

describe('Button', () => {
  test('must return the component template', () => {
    render(<Button data-testid='button' id='button_id' className='temp_class' />)

    expect(screen.getByTestId('button')).toMatchSnapshot()
  })

  test('must render startIcon', () => {
    render(<Button startIcon={() => <i data-testid='button_icon' />} />)

    expect(screen.getByTestId('button_icon')).toBeInTheDocument()
  })
  test('must render endIcon', () => {
    render(<Button endIcon={() => <i data-testid='button_icon' />} />)

    expect(screen.getByTestId('button_icon')).toBeInTheDocument()
  })

  test('must render BUTTON tag', () => {
    render(<Button data-testid='button' />)

    expect(screen.getByTestId('button')).toHaveAttribute('type', 'button')
  })

  test('must render children', () => {
    render(<Button>children</Button>)

    expect(screen.getByRole('button')).toHaveTextContent('children')
  })

  test('must call the function when Button clicked', () => {
    const onClick = jest.fn()
    render(<Button onClick={onClick} />)

    expect(onClick.mock.calls.length).toEqual(0)
    fireEvent.click(screen.getByRole('button'))
    expect(onClick.mock.calls.length).toEqual(1)
  })
})
