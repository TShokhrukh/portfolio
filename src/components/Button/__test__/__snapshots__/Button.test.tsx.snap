// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button must return the component template 1`] = `
<button
  class="Button temp_class"
  data-testid="button"
  id="button_id"
  type="button"
/>
`;
